src/Hamiltonians/Molecular.jl

```diff
@@ -0,0 +1,201 @@
+"""
+    MolecularMolecularHamiltonian(fcidump::String)
+    
+    MolecularMolecularHamiltonian(fd::ElemCo.FciDumps.FDump)
+
+Implements a Molecular Hamiltonian with ElemCo.jl.
+"""
+struct MolecularHamiltonian{T,A<:FermiFS2C,D<:FDump} <: AbstractHamiltonian{T}
+    starting_address::A
+    fcidump::D
+end
+
+function MolecularHamiltonian(fcidump::String)
+    fd = read_fcidump(fcidump, Val(4))
+    MolecularHamiltonian(fd)
+end
+
+function MolecularHamiltonian(fd::QFDump)
+    n_orb = headvar(fd, "NORB")
+    n_elec = headvar(fd, "NELEC")
+    ms2 = headvar(fd, "MS2")
+
+    n_alpha_elec = (n_elec + ms2) ÷ 2
+    n_beta_elec = (n_elec - ms2) ÷ 2
+
+    starting_address = FermiFS2C(near_uniform(FermiFS{n_alpha_elec,n_orb}), near_uniform(FermiFS{n_beta_elec,n_orb}))
+
+    val_type = typeof(fd.int0)
+    addr_type = typeof(starting_address)
+    fd_type = typeof(fd)
+    MolecularHamiltonian{val_type,addr_type,fd_type}(starting_address, fd)
+end
+
+struct Modes{T<:FermiFSIndex}
+    occupied::Vector{Vector{T}}
+    unoccupied::Vector{Vector{T}}
+end
+
+function modes_extract(addr::FermiFS2C)
+    occupied_modes = Vector{Vector{FermiFSIndex}}(undef, 2)
+    unoccupied_modes = Vector{Vector{FermiFSIndex}}(undef, 2)
+    for (i, state) in enumerate(addr.components)
+        occupied_modes[i] = OccupiedModeMap(state)
+        state_unoccupied_modes = FermiFSIndex[]
+        for j = 1:num_modes(state)
+            ind = find_mode(state, j)
+            if ind.occnum == 0
+                push!(state_unoccupied_modes, ind)
+            end
+        end
+        unoccupied_modes[i] = state_unoccupied_modes
+    end
+    Modes(occupied_modes, unoccupied_modes)
+end
```

Comment:
As discussed, this look similar to what `OccupiedModeMap` is already doing. It would be better to extend the code in `src/BitstringAddresses/fockaddress.jl` to cater for `CompositeFS` addresses, and possibly for empty modes as well if that is required. It should be done using similar techniques as used for `OccupiedModeMap` including `SVector`s or `Tuple`s rather than `Vector`s to avoid heap allocations.